doc/decisions/header_file_structure.md

```diff
@@ -12,31 +12,135 @@ The big problem is kdbprivate.h. It has two main problems:
+
+- `src/include/elektra/foo/*.h`:
+  Additional public API header of `libelektra-foo`.
+  Will be installed as `<include-root>/elektra/foo/*.h`.
+
+  All of these installed headers `installed.h` must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.
+
+- `src/include/internal/foo.h`:
+  Declares the internal API of `libelektra-foo`, by including `#include <internal/foo/*.h>`.
+  Will not be installed.
+
+  Such a header may only contain `#include <internal/foo/*.h>` lines.
+
+- `src/include/internal/foo/*.h`:
+  Additional internal header of `libelektra-foo`.
+  Will not be installed.
+
+  All of these installed headers `installed.h` must be included from `src/include/internal/foo.h` via a line `#include <internal/foo/installed.h>`.
+
+Additionally, all libraries may also have private headers:
+
+- `src/lib/foo/**/*.h`:
+  Additional headers may be present.
+  These headers may only be used by other files within `src/lib/foo`, according to the rules in [Including Headers](header_include.md).
+
+### Plugins
+
+Plugins do not declare their API via header files.
+Their headers are never installed and can be named any way the developer wants.
+
+### Tests
+
+For category 2 from above (private but needs to be tested), one of the following should be done:
+
+1. Declare functions as `static` in a `.c` file and `#include ""` this file from the test.
+2. Add a private non-installed header (e.g., `src/lib/foo/testapi.h`) that declares the API that needs testing, `#include ""` that and compile the test sources together with the `.o` files from the library (static linking).
+
+If a symbol is needed in only one file and for tests, option 1 should be preferred.
+For symbols that are used in multiple files, a header needs to exist anyway.
+In any case, these unit tests should not be installed and should statically link the library into the test executable.
+This way we don't pollute our `symbols.map` files and keep the number of exported symbols down.
 
 ## Rationale
 
+- This structure makes the `#include`s simple and works nicely with our directory structure.
+- Having a uniform naming convention simplifies things, both for developers writing Elektra and those using Elektra.
+- Requiring libraries must either be fully modularized (main headers are only `#includes`) or completely monolithic, avoids the situation where a library has some parts in `foo.h` and other parts in extra headers.
+  This is bad, because it encourages users to simply include `foo.h`, which defeats the point of modularized headers.
+
 ## Implications
 
+- The location of a header file within the source tree determines what API it contains:
+  - Public: Will be installed, must be stable, can be used anywhere
+  - Internal: Will not be installed, doesn't have to be stable, can be used anywhere, but public headers
+  - Private: Will not be installed, doesn't have to be stable, can only be used in the same library
```

Comment:
This private is different to the 4 private above?

Reply:
It's category 1. Truly private shouldn't be used outside library.